Get more fusion after autodiff uses SumToSize

t-vi opened the pull request:
Here is a fresh attempt at getting some fusion back in autodiff-generated graphs in the presence of SumToSize.

- The sum to size operator is now  `aten::_grad_sum_to_size` to allow symbolic script differentiation (and that in turn would need to use this in place of sum_to_size to signal that it strictly operates on gradients). This is also used in the autodiff code, replacing `prim::SumToSize`.
- `_grad_sum_to_size` is now fusable, `cat`s - which are fused afterwards thanks to Adam's simplification of the code - are only fused if there is no `_grad_sum_to_size` in the fusion group.
- I push the `_grad_sum_to_size` out of the the fusion group when compiling and record the desired summations in the KernelSpec. The reasoning is the following:
  - As the autodiff is a repeated applicaiton of the chain rule, we always have the pattern `grad_in = mm(A, grad_out)`,  with A often diagonal for cases interesting to the fuser, whence it is `grad_in = a * grad_out` (a pointwise multiplication). We know that only `grad_out` may have AutodiffGradSumToSize applied, so we can commute AutodiffGradSumToSize with the `mul` (and `div` and `neg` are of similar origin).
  - For `type_as` the gradient might be giving the type, so just skip SumToSize,
  - `add` (which was inserted as `prim::AutogradAdd`) adding gradients when the forward used the same value in several places. This is non-broadcasting, so we know that the two arguments would have the same sizes as inputs - which is good so we don't have to do bookkeeping of the two parts.

Details:
- During fusion, the Tensor arguments are always kept as the first parameters of the fusion group to accomodate indexing assumptions in the fuser.
- The rewriting of the fusion group to record the necessary output transformation and eliminate `_grad_sum_to_size` from the fusion group is now in the fuser compile step.
- In the execution step, the arguments are split into Tensor / Non-Tensor and the non-tensor args are mostly forgotten about except for doing `sum_to_size` at the end. This would want to be improved if/when we fuse nonconstant scalar arguments.
- In a number of places in the fuser, the non-Tensor arguments to the fusion group needed to be ignored.

Thank you, @apaszke for the insightful discussion. All bad ideas and errors are my own.

t-vi committed:
start on re-fusing backwards

t-vi committed:
Merge branch 'master' into fusion_after_autodiff

t-vi committed:
fix typo

t-vi committed:
matching, update tests

apaszke commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -79,7 +79,7 @@ bool isDifferentiable(Node * n) {
     "aten::tan(Tensor self) -> Tensor",
     "aten::trunc(Tensor self) -> Tensor",
-    "prim::SumToSize(Tensor(a) self, int[] size) -> Tensor(a)",
+    "prim::AutodiffGradSumToSize(Tensor(a) self, int[] size) -> Tensor(a)",
```
I think GradSumToSize would be enough 😄

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -138,7 +138,9 @@ struct GraphFuser {
     if (node->owningBlock() != block_) return false;
-    return node->kind() == prim::FusionGroup || isSimpleMap(node);
+    return node->kind() == prim::FusionGroup
+      || node->kind() == prim::AutodiffGradSumToSize
```
Um this is untrue. You can’t always fuse those nodes.

t-vi replied:
Fair enough, even if the logic now is a lot more complicated than I would like (especially given that we expect to be able to fuse them again).

apaszke reviewed (commented):
I think the removal of SumToSize nodes is happening way too early. Basically, you shouldn’t think of FusionGroups as graphs that have already been fused and will conform to those semantics, but graphs eligible to be fused. That means that we still want to preserve the original semantics of the code, because it might turn out that our fusion guesses were wrong, and will end up running a deoptimized version of the original code. Instead, we should allow putting them in FusionGroups and simply remove them right before a kernel is compiled (once we know that the fusion is valid, etc.).

Finally, marking those nodes as fusible is a bad idea, because the only reason why you might put them in a fusion group is because you are certain that it will help you perform more fusions. That should be checked and processed similarly to how we deal with rearranging chunk nodes.

t-vi commented:
Thanks for your comments Adam!

- I'll rename the `prim::GradSumToSize`.
- So I'll move the graph rewriting into the fuser codegen.
- For "when to fuse SumToSize", would it be OK to put them into the fusion group if there aren't any FusedConcat nodes in there? 
  This would mean that we might end up with SumToSize at the top of the fusion group which we would undo before the fusion, but I'm a bit weary that `GraphFuser.run` will get considerably more complicated if we split out the scan phase as done for `chunk`.

t-vi committed:
it's GradSumToSize

apaszke committed:
Add support for batch_norm fusion to the JIT

We don't support reductions yet, but simply decomposing batch_norm
into a kernel that computes the stats, and the fusing everything else
with ReLU and following pointwise ops provides nice speedups.

Note that this is only limited to inference mode for now, because we
don't support convolutions and batch norm in AD, so the fuser isn't
applied to those parts.

apaszke committed:
Add support for batch_norm fusion to the JIT

We don't support reductions yet, but simply decomposing batch_norm
into a kernel that computes the stats, and the fusing everything else
with ReLU and following pointwise ops provides nice speedups.

Note that this is only limited to inference mode for now, because we
don't support convolutions and batch norm in AD, so the fuser isn't
applied to those parts.

t-vi committed:
improve detection of fusible GradSumToSize

t-vi committed:
fuse less GradSumToSize

t-vi commented:
Hmh. I need to rebase.
So I think I'm not fusing sumtosize any more when concat is close.
I'm not as sure about the "when to relocate sumtosize": If I move that into kernel generation, is it still safe to move the sumtosize to outside the fusion group? I'll try that next, but I'm still a bit sceptical about it.

apaszke committed:
clang-format complains...

apaszke committed:
clang-format complains...

apaszke committed:
Adjust precision

apaszke committed:
Adjust precision

apaszke committed:
Separate aten::cat fusion from pointwise fusion

That makes that definition of a "fusable node" much simpler,
as we don't need to keep considering whether something has to be an
"exit node" at every step. The fuser now tries to maximize the
pointwise fusions first, and proceeds to prepending chunks and appending
concats only once a fix point is reached.

This patch not only makes the fuser much simpler to reason about,
making it siginifcantly easier to implement features like SumToSize
fusion, to improve performance of derivative graphs.

apaszke committed:
Separate aten::cat fusion from pointwise fusion

That makes that definition of a "fusable node" much simpler,
as we don't need to keep considering whether something has to be an
"exit node" at every step. The fuser now tries to maximize the
pointwise fusions first, and proceeds to prepending chunks and appending
concats only once a fix point is reached.

This patch not only makes the fuser much simpler to reason about,
making it siginifcantly easier to implement features like SumToSize
fusion, to improve performance of derivative graphs.

apaszke committed:
Implement batch_norm_update_stats for CPU

mruberry committed:
removes shape logic from code generation, fixes possible segfault

mruberry committed:
cleanup

mruberry committed:
moves rand review to spec constructor to initialize new field

apaszke committed:
Review comments

t-vi committed:
Merge remote-tracking branch 'apaszke/bn_fusion' into simplify_cat_fusion

t-vi committed:
Merge branch 'simplify_cat_fusion' into fusion_after_autodiff

t-vi committed:
Merge branch 'pr_15750' into fusion_after_autodiff

t-vi committed:
update after merge

apaszke committed:
Review comments

t-vi committed:
more fusion for backwards v2

t-vi committed:
Merge branch 'simplify_cat_fusion' into fusion_after_autodiff

t-vi committed:
update expect

t-vi commented:
So now I get a fully fused backward for the IoU-loss again.
The PR is based on #15633 (simplify cat fusion), which in turn depends on #15897 (batch norm fusion). The latter was merged and reverted yesterday.

I'm not quite sure how to rearrange the PR commits to facilitate review, if you have a preference, let me know. I'm leaning towards squashing everything (beyond #15633) into a large commit.

I'll edit the start comment to describe the PR a bit more.

t-vi committed:
Merge branch 'master' into fusion_after_autodiff

zou3519 commented on test/test_jit.py:
```diff
@@ -10356,7 +10356,7 @@ def funcInf(a, b):
             c.sum().backward()
             graph = backward_graph(s)
-            self.assertAllFused(graph, except_for={'prim::SumToSize'})
+            self.assertAllFused(graph, except_for={'prim::GradSumToSize'})
```
is this supposed to be `aten::_grad_sum_to_size`?

zou3519 commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -1009,7 +1009,7 @@ static void eliminateDeadCode(ReverseDetails& rev_info) {
 static void Optimize(Gradient& grad_desc, ReverseDetails& rev_info) {
-  // TODO: we are sometimes emitting expressions like SumToSize(SumToSize(x,
+  // TODO: we are sometimes emitting expressions like GradSumToSize(GradSumToSize(x,
   // s1), s2), which are equivalent to SumToSize(x, s2), and could save us some
```
Should update this line of the comment too

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -86,11 +87,13 @@ static const Node* usedInFusedChunk(const Value* input) {
-          chunk->i(attr::chunks), chunk->i(attr::dim));
-    } else {
-      spec.inputChunks().emplace_back(1, 0);
+    if (input->type()->isSubtypeOf(DynamicType::get())) {
```
nit: 
```
if (!input->type()->isSubtypeOf(DynamicType::get())) {
   continue;
}
```
keeps the blame

t-vi replied:
I must admit that `continue` is very foreign to me (feels like goto), but that's me and I'll adapt.

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -104,7 +107,8 @@ static std::vector<int64_t> getInputDependencies(const Value* output) {
     queue.pop_back();
     const Node* producer = val->node();
-    if (producer->kind() == prim::Param) {
+    if (producer->kind() == prim::Param &&
```
Maybe add a big comment here that non-inputs must be sizes that are not used for the output computation (and only used for gradient computation)

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +148,85 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
 }
 
+// This function moves GradSumToSize nodes inside a fusion group to the end of
+// the fusion group. Note that correctness relies on the invariant that
```
> to the end of the fusion group

Are they being moved outside the fusion group?

t-vi replied:
They're eliminated by pushing them to the outputs of the fusion group, recording their effect on the outputs, and deleting them. I'll update the comment.

zou3519 commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -179,6 +179,16 @@ struct GraphFuser {
 
+  bool containsGradSumToSize(Node* fusion_group) {
+    auto subgraph = &getSubgraph(fusion_group);
+    for (Node* n : subgraph->nodes()) {
```
nit: it's a little more concise with std::find_if

zou3519 commented on torch/csrc/jit/fuser/kernel_spec.h:
```diff
@@ -127,6 +132,7 @@ struct TORCH_API KernelSpec {
   uint64_t nInputs_;
   std::vector<std::vector<int64_t>> inputBroadcastGroups_;
   std::vector<PartitionInfo> inputChunks_;
+  std::vector<size_t> outputGradSumToSizes_;
```
What is the `size_t`? An offset?

t-vi replied:
Oh no. :( Something that is an index but can be -1 to say "None". So it should be int64_t probably.

zou3519 replied:
ah maybe `at::optional<size_t>` would work then... (or just `int64_t` as you're proposing)

t-vi replied:
I prefer int64_t if I may, but if you prefer at:optional, I'd go with that.

zou3519 replied:
No preferences on my part, int64_t is fine

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +148,85 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+  // Scan the graph. As we will delete nodes, we use the backward ordering.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend();) {
+    auto* node = *it;
+    ++it; // We delete the GradSumToSize nodes, so increment the iterator now.
```
There's an `it.destoryCurrent()` that might work here

t-vi replied:
But so I would do `it.destroyCurrent(); it++` instead of `it++` and then `node->destroy()`?

zou3519 replied:
I think instead of `node->destroy()` you can just use `it.destoryCurrent()`

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +148,85 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+          if (outputGradSumToSizes[offset] == -1) {
+            // note: we make the assumption that the sizes are inputs to the
+            // fusion group (rather than something calculated).
+            outputGradSumToSizes[offset] = node->inputs()[1]->offset();
```
Is there an assumption that if the input `x` to prim::Return doesn't come from a _grad_sum_to_sizes node, then nowhere in the computation a _grad_sum_to_sizes node is used to compute `x`?

t-vi replied:
No, the output of `_grad_sum_to_sizes` is  "tracked" to the output. If it isn't found at all, then yes, it's not summed. The implicit assumption here is that if several `_grad_sum_to_sizes` - so the `outputGradSumToSize[offset]` has been set to something, it doesn't matter which one we use.

t-vi replied:
I amended the commentary, but somehow github doesn't pick it up to mark this outdated.

zou3519 commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +148,85 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // we can just keep the last, so we're done here (this case might not
+          // happen due to how we organize the search and remove nodes...)
```
Well, because we're iterating in reverse topological order, it sounds like this shouldn't happen. It might be better to `JIT_ASSERT(false)` here?

t-vi replied:
OK, I'll try that.

zou3519 commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -350,16 +361,28 @@ struct GraphFuser {
     // add n's inputs to the fusion group's input list if we don't already have
     // them
+    // we insert tensors first because the fuser assumes that to be the case
+    // (as a legacy from tensors only)
```
Observation: it's not the nicest thing in the world to say that all tensor arguments should come before the non-tensor arguments, but I don't know if the alternative is worse. If we stick with this approach we should rename `mergeNodeIntoGroup` to reflect how the argument list contains tensors first and non-tensors later.

t-vi replied:
The problem is that the fuser computes several vectors (and indices into them) that are only valid for tensor inputs. I first tried to use argument indices without reordering, but it's fairly hairy to keep track of, so I figured that it wasn't worth it. I briefly discussed that with @apaszke and he thought that it wasn't too bad to reorder the fusion group inputs. Later on I decided that reordering was more trouble than enforcing the constraint from the beginning.
Long story short, someone more knowledgeable or smarter than me would probably not need the constraint, but I got lost in the required bookkeeping

ngimel commented on torch/csrc/jit/fuser/executor.cpp:
```diff
@@ -339,6 +344,12 @@ bool runFusion(const int64_t key, Stack& stack) {
   std::vector<at::Tensor> outputs;
   launchFusion(*(*maybe_kernel), device, inputs, outputs);
 
+  for (size_t i = 0; i < outputs.size(); i++) {
```
I don't know if it affects perf a lot, but at this point with runtime shapes information you already know when in fact there are no reductions required, so this at::sum_to can be skipped altogether.

t-vi replied:
I would hope that sum_to is fast if there isn't anything to be done, even if it compares number of tensors * average dimension sizes.

What would be the condition to skip the entire thing?

ngimel replied:
It should be fast, it's just I've been seeing cases lately where CPU cannot get ahead of GPU and all 20-50 us latencies arising even in cheap ops are exposed. I think sufficient condition is if no argument was expanded https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/fuser/executor.cpp#L110

t-vi replied:
As far as I understood, this is hard to skip.

ngimel commented:
@t-vi, with this PR applied, the generated backward kernel for the following function
```
def fn1(x,y,z):
    a = x+y+z     
    return torch.sigmoid(a)
```
separately materializes gradients to x,y and z
```
float n0 = t0.data[t0_offset];
      float n1 = t1.data[t1_offset];
      int n5 = 1;
      float n6 = -n1;
      float n7 = n6 + ((float) n5)*((float) n5);
      float n8 = n7 * n1;
      float n9 = n8 * n0;
      t2.data[t2_offset] = n9;
      t3.data[t3_offset] = n9;
      t4.data[t4_offset] = n9;
```
In current master without this PR fusion group has a single output, that later serves as an input to 3 SumToSize ops.

t-vi committed:
Address first round of review comments

Thank you, @zou3519 !

t-vi commented:
Thanks @ngimel, for raising this. I'll see how to fix that. My understanding is that we would need to deduplicate it for the kernel, but not for the _grad_sum_to_size application after running the fused kernel. That in turn means we have different outputs for the fused kernel vs. the fusion group.

t-vi committed:
Deduplicate fusion kernel outputs

After squeezing out the _grad_sum_to_size during kernel compilation,
we may end up with duplicate outputs.
For example example, the backward of
    def fn1(x,y,z):
        a = x+y+z
        return torch.sigmoid(a)
has that.
Thank you @ngimel for noting and providing the example!

t-vi committed:
add test for fuser deduplication

t-vi committed:
clang-tidy

t-vi commented:
So I added the output deduplication in the fuser and a test using @ngimel 's example (thanks again!).

ngimel commented on test/test_jit.py:
```diff
@@ -10512,6 +10512,26 @@ def f(x):
+        graph = backward_graph(s)
+        self.assertAllFused(graph)
+        # check that a, b share storage, i.e. were generated as a single output in the fuser
+        self.assertEqual(ga, gb)
```
I thought this would check that contents of the tensors is the same, but not that they share storage?

t-vi replied:
Oh no. I forgot the `.data_ptr()`. Thanks for catching this.

t-vi committed:
add .data_ptr() to check equal storage

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -188,7 +195,8 @@ struct GraphFuser {
       return false;
-    return node->kind() == prim::FusionGroup || isSimpleMap(node);
+    return node->kind() == prim::FusionGroup ||
+        node->kind() == aten::_grad_sum_to_size || isSimpleMap(node);
```
I'd really like us to avoid putting `_grad_sum_to_size` nodes in fusion groups unnecessarily. We should either have stronger checks for them (that adding them would in fact help us fuse more), or we should have a postprocessing pass that e.g. will move them out of the group if they are applied to inputs, or create outputs.

t-vi replied:
Thanks for going through this!

So postprocessing or checking if the input of _grad_sum_to_size is fusible? Do you have a preference?
Would the "create outputs" comment be mitigated by the deduplication in the fuser itself?
Personally, I would envision that one would move cases where the sumtosizes of the outputs are "ascending" (i.e. you can sort the dimensions in a way that every tensor only has the summations last) into the kernel itself.

t-vi replied:
So I now do this. This probably means we do not fuse some cases where we would like to - e.g. the milstm backward could be such a case, but we are doing better than before and maybe we just are conservative.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -917,7 +938,8 @@ struct GraphFuser {
     JIT_ASSERT(inputs.size() == sinputs.size());
     for (size_t i = 0; i < inputs.size(); ++i) {
-      shape_of[sinputs[i]] = graph->insert(aten::size, {inputs[i]});
+      if (inputs[i]->type()->isSubtypeOf(DynamicType::get()))
```
nit: please add braces. This is somewhat hard to read

apaszke commented on torch/csrc/jit/fuser/kernel_spec.h:
```diff
@@ -41,6 +41,24 @@ struct TORCH_API PartitionInfo {
   int64_t dim_;
 };
 
+struct TORCH_API OutputMapAndSize {
```
This badly needs docs

apaszke commented on torch/csrc/jit/fuser/kernel_spec.h:
```diff
@@ -127,6 +150,7 @@ struct TORCH_API KernelSpec {
   uint64_t nInputs_;
   std::vector<std::vector<int64_t>> inputBroadcastGroups_;
   std::vector<PartitionInfo> inputChunks_;
+  std::vector<OutputMapAndSize> outputMapAndSizes_;
```
This needs docs too. How many of elements does this have? As many as there are tensor inputs?

apaszke commented on torch/csrc/jit/fuser/executor.cpp:
```diff
@@ -297,9 +297,14 @@ bool runFusion(const int64_t key, Stack& stack) {
+  std::vector<at::Tensor> inputs;
+  inputs.reserve(all_inputs.size());
+  for (auto& i : all_inputs) {
+    if (i.isTensor()) {
```
Can't we simply have an invariant that tensor inputs precede other inputs? Then, you could compute the length of this loop once, and avoid unnecessary checks on size inputs.

t-vi replied:
:+1:  We already have that invariant.

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -86,6 +86,9 @@ static const Node* usedInFusedChunk(const Value* input) {
   for (const Value* input : (spec.graph())->inputs()) {
+    if (!input->type()->isSubtypeOf(DynamicType::get())) {
+      continue;
+    }
```
Please document the fact that we only have as many chunk descriptors as tensor inputs. Those things were simple previously, but are getting complicated now.

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+// Note that the correctness relies on the invariant that
+// _grad_sum_to_size is only applied to gradient nodes created by autodiff.
+// This is important because it ensures that in the mul and div nodes only
+// one argument (in the case of diff the numerator) has a summed value.
```
You probably wanted to say "in case of _div_ the numerator"

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -206,16 +324,18 @@ std::shared_ptr<FusedKernel> compileKernel(
-        chunk_desc.emplace_back(input_desc[input_index++], chunks, dim);
-        for (const auto* o : chunk->outputs()) {
-          flat_inputs.emplace_back(o, *chunk_desc.back().subTensorDesc());
+      if (p->type()->isSubtypeOf(DynamicType::get())) {
```
nit: a simpler way would be to do 
```cpp
if (!p->type()->isSubtypeOf(DynamicType::get())) {
  continue;
}
```

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+void processGradSumToSize(KernelSpec& spec) {
+  auto graph = spec.graph();
+
+  // keep track of one sumToSize hitting each output
```
nit: "of _the last_ sumToSize on the path to an output". There's _at most_ one, so this sentence might not be true

apaszke replied:
another nit -- this is not used until the very end of this function. Maybe move it there?

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
```
Ugh! Please reverse them first and then iterate in order, or this becomes quadratic in the number of uses!!

apaszke replied:
Also, I don't get why would you do that, since uses are not sorted topologically. If that's what you're after, you should sort them explicitly (see the fuser for code that can do that).

apaszke replied:
Ok, reading your code later it seems that you simply want a FIFO queue! In that case, use `std::queue`, which will give you O(1) complexity.

t-vi replied:
`std::queue<Use>` or `std::queue<Use*>`? I don't have any luck with using `std::queue<Use&>`.

apaszke replied:
`std::queue<Use>` is fine. It's a 16B struct.

apaszke replied:
Also, it should be `Use` because `Use*` will get invalidate in case you modify them.

t-vi replied:
Them I'm glad that I picked the right thing. :)

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+  // we scan backwards.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend(); it++) {
+    auto* node = *it;
+    if (node->kind() == aten::_grad_sum_to_size) {
```
Can we just do this?

```cpp
if (!node->kind() == aten::_grad_sum_to_size) {
  continue;
}
```

This has both the benefit of decreasing indentation, and makes it clear that it's the only case you want to handle here (I had to scroll down 50 lines just to find out that all elses later don't apply to this if).

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // this case should not happen due to how we organize the search
+          // and remove nodes
```
It would be much simpler to explicitly say "This can never happen because we iterate backwards and delete all _sum_to_size nodes we see".

apaszke replied:
Also, I kind of feel like it would be nicer to put this assertion before this `if ... else` chain

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+                "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
+          // sometimes, a mask or similar is cast to the same
+          // type as the gradient. But we only want to allow that as offset==1
+          // (i.e. the type)
```
Sorry, why do we only want to allow the output of `_grad_sum_to_size` to be used as a template for a cast and not as `self`? That seems like a too strong condition for an assert.

t-vi replied:
Indeed, I'll absorb sumtosize hitting other and pass through self.

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+          // on the gradient (matmul would likely be, too but we don't fuse it)
+          // note that for mul, we know (from the chain rule) that only one
+          // factor will be stemming from a calculation involving gradients so
+          // we know that we can move SumToSize across it
```
This comment should really live next to the definition of `commutes_with_SumToSize`

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+        } else {
+          throw std::runtime_error(
+              "unknown node type for fusion SumToSize processing");
+        }
```
I feel like this is a bit too strong of a condition. `commutes_with_SumToSize` doesn't exhaust all fusable nodes, so in fact you _can_ trigger this case, and what then? We can't land a patch that has no fallback for such a case.

apaszke replied:
That's a pretty strong assumption to be honest. We should still guard against it to make it more future proof. I wouldn't be surprised if we wanted to emit sth like `at::_sigmoid_backward` to use the optimized implementation if it doesn't get fused.

t-vi replied:
Now we fuse only if we can deal with the processing needed, so I think that's cool now.

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+            // fusion group (rather than something calculated).
+            outputGradSumToSizes[offset] = node->inputs()[1]->offset();
+          }
+        } else if (user->kind() == prim::FusedConcat) {
```
You could simply assert that this never happens (by design)

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+          // because we want later (in the graph) nodes to be preferrable
+          // to earlier ones and we iterate backwards
+          // The implicit assumption is that if we have several _grad_sumtosizes
+          // "in parallel" (from addition) they are the same.
```
I don't get the part about "from addition". Can you please elaborate?

t-vi replied:
If you have the backward of `y = torch.sin(phi)**2 + torch.cos(phi)**2`, you'll have `dy/dphi = something_from_sin + something_from_cos`. Now if you do different broadcasting or do something where the fusion group has several outputs, you have several inputs. However, given that the final sumtosize would be to the shape of phi, they should all be the same. (Of course, in practice, the gradient might not always as easy to compute as in the above example.)

apaszke commented on torch/csrc/jit/fuser/compiler.cpp:
```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+      graph->eraseOutput(newo);
+      outputMapAndSizes.emplace_back(it->second, osize);
+    }
+  }
```
Ohhh interesting. So `outputMapAndSizes` really describes the postprocessing we do on the kernel outputs, and we might duplicate them! That should really go in the some comments.

apaszke reviewed (changes requested):
I think this should still be improved before we land. First, it can fail when one encounters some non-basic fused nodes, and there's no fallback in that case. Also, it would be good to improve the quality of fusability checks for `aten::_grad_sum_to_size` to avoid fusing them unnecessarily.

t-vi committed:
Improve many bits based on Adam's feedback.

Thank you!

t-vi committed:
update milstm backward expectation to reflect less fusion

t-vi committed:
Merge branch 'master' into fusion_after_autodiff

t-vi committed:
follow JIT_ASSERT->AT_ASSERT

t-vi commented:
@apaszke I think I acted on your comments (though I have the feeling that there might be a nicer way to cover the two uses of `trackSingleGradSumToSizeToOutputs`, but I don't really have an idea how).

t-vi committed:
import queue

t-vi committed:
python 2

t-vi committed:
python2 ?!

t-vi committed:
Merge branch 'master' into fusion_after_autodiff

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -1202,6 +1249,78 @@ void PeepholeOptimizeShapeExpressions(Block* block) {
+// will be recorded as target sizes for the outputs as applicable.
+// In the graph_fuser pass we only need to check that we can go to the
+// outputs while in the fuser's compiler we want to record the sizes.
+bool trackSingleGradSumToSizeToOutputs(
```
It's important to mention that those calls should be done in *reverse* order w.r.t. the order of the resulting fusion group

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -1202,6 +1249,78 @@ void PeepholeOptimizeShapeExpressions(Block* block) {
+// outputs while in the fuser's compiler we want to record the sizes.
+bool trackSingleGradSumToSizeToOutputs(
+    Value* gradSumToSizeOutput,
+    c10::optional<std::vector<int64_t>*> outputGradSumToSizes) {
```
What's the point of taking an optional pointer? Can't you just use `nullptr` as the default value?

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -1202,6 +1249,78 @@ void PeepholeOptimizeShapeExpressions(Block* block) {
+      // do nothing
+      // this case only happens in the graph_fuser step because in the
+      // compile stap because we iterate backwards and delete
+      // all _grad_sum__to_size nodes we see
```
nit: small typos in here: stap, _grad_sum__to_size

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -188,6 +196,13 @@ struct GraphFuser {
+      // only fuse _grad_sum_to_size if
+      // - we will fuse its input next
+      // - we can commute the _grad_sum_to_size with everything
+      //   along the computation graph until we reach the outputs
```
It would be good to say that this second condition will only be checked later.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -444,6 +472,21 @@ struct GraphFuser {
       return at::nullopt;
     }
+    if (producer->node()->kind() == aten::_grad_sum_to_size &&
+	consumer->kind() == prim::FusionGroup) {
```
Can we assert that `consumer` is a fusion group? If I understand correctly this is the only case possible by design, but having a check like this would mean that this case will simply get skipped if it wasn't the case.

t-vi replied:
I think the case where the consumer isn't a fusion group, i.e. a new fusion group will be started with a `_grad_sum_to_size` for the output, is legitimate, and then we would have the empty set to check.
The main alternative I see would be to refuse to fuse here, but that would only result in us having two _grad_sum_to_sizes in a row (one pushed from inside the fusion group and one that we didn't fuse because it was at the output) during execution.

apaszke reviewed (approved):
Looks great! Some minor comments, but should be good to land

t-vi commented on test/test_jit.py:
```diff
@@ -10947,6 +10947,68 @@ def f(x):
+            rbx = torch.min(b1x2, b2x2)
+            rby = torch.min(b1y2, b2y2)
+
+            TO_REMOVE = 1
```
It's going to be removed. :)

t-vi committed:
Second round of Adam's feedback. Thank you again!

t-vi commented:
So the three CI failures would appear to be unrelated to my changes (flake8 in onnx/symbolic.py which I don't think I touch, python to old to download mypy, UtilsNMSTest.GPUEqualsCPUCorrectnessTest - this is perhaps the one I know the least about).

apaszke commented:
This is exciting! I'll have to rerun my benchmarks now!
